### PR TITLE
Update django-phonenumber-field dependency to newer version

### PR DESCRIPTION
**Pull request by syserr0r**
Untested attempt to allow newer django-phonenumber-field dependency for issue #387

## Description
As above
PS: I would like to know why you required <3.99 and not <4 -- is this best practice or for some other reason?

## Motivation and Context
Issue #387

## How Has This Been Tested?
It hasn't.
https://github.com/stefanfoulis/django-phonenumber-field/blob/master/CHANGELOG.rst reports no breaking changes so fingers crossed, eh?

## Screenshots (if appropriate):
N/a

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
  - Provide Django 3.1 support in dependency
  - Allow package use when people are already using django-phonenumber-field 4 or newer
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by moggers87** [comment]
No, "fingers crossed" is not good enough. Actual user testing is required, i.e. install 5.0.0 and see if you can still set a phone number.

**Comment by claudep** [comment]
Shouldn't `https://github.com/Bouke/django-two-factor-auth/blob/3a0fdfa191a64ded99c607e9040197803cee3f87/tests/test_views_phone.py` catch any errors in case of new incompatibilities?

**Comment by moggers87** [comment]
Not sure, I wouldn't rely on it alone :smile_cat: 

There's also the changes in 4.0.0 that were discussed in https://github.com/Bouke/django-two-factor-auth/pull/326. As I understand it, that would allow invalid phone numbers that would have been previously rejected. Do our unit tests cover the case where Twilio rejects a number as invalid or were we relying on django-phonenumbers-field's "very aggressive validation". I suspect we'd be fine, but I'd rather be safe than sorry.

**Comment by claudep** [comment]
Unfortunately, I'm not using phone capability in my projects at all, so I'm not in a good position for testing.

**Comment by syserr0r** [comment]
> No, "fingers crossed" is not good enough. Actual user testing is required, i.e. install 5.0.0 and see if you can still set a phone number.

Sorry, just assumed when you released an update supporting Django 3.1 but your dependency doesn't that you had enough test coverage to catch issues.

If I get a chance I'll see if I can run it on our test server.

**Comment by syserr0r** [comment]
Looks like we've been running 5.0.0 in staging and production for a while -- the old pip resolver warned about it but installed both anyway:

From the build log:
```
django-two-factor-auth 1.12.1 requires django-phonenumber-field<3.99,>=1.1.0, but you'll have django-phonenumber-field 5.0.0 which is incompatible.
```

From a node:
```
root@production-c88989d7d-rwb5k:/src# pip freeze | grep  -E '(two-factor|phonenumber-field)'
django-phonenumber-field==5.0.0
django-two-factor-auth==1.12.1
```

What outliers would you have me test?

**Comment by kevindice** [comment]
Can confirm, 5.0.0 works in production w/ 1.13.0.

Now, since 5.0.0 removes support for the end-of-life Django 1.11 and 2.2, would it make sense for the purposes of larger project planing (@moggers87 ?) to slot this in for a major version bump of `django-two-factor-auth`?

I'm seeing that Django 1.11 support was dropped in `django-two-factor-auth` version 1.12.

A couple options:
- `django-two-factor-auth` drops support for end-of-life Django 2.2, changes the requirement to `'django-phonenumber-field>=5,<6'`, and (just going off of convention here) bumps its own version to 1.14.0 to signal a breaking change (loss of 2.2 support)
- `django-two-factor-auth` preserves official support for Django 2.2 by setting the range to `django-phonenumber-field>=4,<5` (or can allow v5 with it being the user's responsibility to pin a pre-v5 version if they really need Django 2.2 support.)

I'd propose the first option.

It appears that `django-phonenumber-field` has only dropped official support for Django 2.2, but it lives on in the test matrix: https://github.com/stefanfoulis/django-phonenumber-field/blob/master/tox.ini#L16-L17

----


tl;dr:  It'd be fine to set this to >=5 and <6 and release 1.14.0

**Comment by kevindice** [comment]
More on that v4 change - `django-phonenumber-field` added aggressive validation in v2.4.0 and v4 removes this in favor of detecting invalidity in....the `str()` method.

This package isn't concerned with capturing that exception, so there's no code for us to remove there.

This issue has a bit more urgency now that pip has made package version conflicts a fatal error as of Dec 1.  That said, it's not that urgent...enjoy your holidays everyone.

If anyone needs a temp fix for your package.json: `git+git://github.com/WeVidIt/django-two-factor-auth@6a4c24b` works

**Comment by moggers87** [comment]
> Now, since 5.0.0 removes support for the end-of-life Django 1.11 and 2.2, would it make sense for the purposes of larger project planing (@moggers87 ?) to slot this in for a major version bump of `django-two-factor-auth`?

This has already been resolved in #387. It was a mistake in the changelog

> This package isn't concerned with capturing that exception, so there's no code for us to remove there.

The concern is phone numbers that would have been rejected by field validation previously will now end up in the database. I've not had a chance to dive into the code to see how well we guard against that when calling Twilio or how well tested that code is.

**Comment by joetheone** [comment]
Is it possible to get an update on this? 🙏

**Comment by moggers87** [comment]
Updated PR title because two people have confirmed it works for them.

Also merged in latest master.

I'll merge this PR if tests pass.

**Comment by atkawa7** [comment]
It seems all tests passes @moggers87. Is there a way for this to get merged

**Review by moggers87 — APPROVED** [review]
LGTM

**Comment by Bouke** [comment]
> I would like to know why you required <3.99 and not <4 -- is this best practice or for some other reason?

When I initially wrote that, `<X` would allow pre-release versions of X. I don't know if that's still the case.
